src/org/jetbrains/plugins/scala/codeInspection/caseClassRedundantNew/RedundantNewCaseClassInspection.scala

```diff
@@ -0,0 +1,92 @@
+import org.jetbrains.plugins.scala.lang.psi.api.expr.ScNewTemplateDefinition
+import org.jetbrains.plugins.scala.lang.psi.api.toplevel.templates.ScClassParents
+import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.ScClass
+import org.jetbrains.plugins.scala.lang.psi.types.api.designator.DesignatorOwner
+import org.jetbrains.plugins.scala.lang.psi.types.{ScParameterizedType, ScType}
+
+/**
+  * mattfowler
+  * 5/7/2016
+  */
+class RedundantNewCaseClassInspection extends AbstractInspection("RedundantNewCaseClass", "Redundant New on Case Class") {
+
+  override def actionFor(holder: ProblemsHolder): PartialFunction[PsiElement, Any] = {
+    case newTemplate: ScNewTemplateDefinition =>
+      if (hasRedundantNew(newTemplate)) {
+        holder.registerProblem(newTemplate.getFirstChild, ScalaBundle.message("new.on.case.class.instantiation.redundant"),
+          ProblemHighlightType.GENERIC_ERROR_OR_WARNING, new RemoveNewQuickFix(newTemplate))
+      }
+  }
+
+  private def hasRedundantNew(newTemplate: ScNewTemplateDefinition): Boolean = {
+    val constructor = getConstructorFromTemplate(newTemplate)
+    val maybeScType: Option[ScType] = getConstructorType(constructor)
+
+    isTypeCaseClass(maybeScType) && isCreatingSameType(newTemplate.getType().toOption, maybeScType) &&
+      constructorCallHasArgumentList(constructor) && isProblemlessPrimaryConstructor(constructor)
+  }
+
+  /**
+    * Determines if the type of the extends block is the same as the type of the new template type.
+    * This prevents us from incorrectly displaying a warning when creating anonymous classes or instances with
+    * mixin traits.
+    */
+  private def isCreatingSameType(newTemplateType: Option[ScType], extendsBlockType: Option[ScType]): Boolean = {
+    (for {
+      templateType <- newTemplateType
+      extendsType <- extendsBlockType
+    } yield {
+      templateType.equals(extendsType)
+    }).getOrElse(false)
+  }
+
+  private def getConstructorFromTemplate(newTemplate: ScNewTemplateDefinition): Option[ScConstructor] = {
+    newTemplate.extendsBlock.firstChild.flatMap {
+      case parents: ScClassParents => parents.constructor
+      case _ => None
+    }
+  }
+
+  private def getConstructorType(maybeConstructor: Option[ScConstructor]): Option[ScType] = {
+    maybeConstructor.flatMap(_.simpleTypeElement.map(_.getType().getOrNothing))
+  }
+
+  private def constructorCallHasArgumentList(maybeConstructor: Option[ScConstructor]): Boolean = {
+    maybeConstructor.flatMap(_.args).isDefined
+  }
+
+  private def isProblemlessPrimaryConstructor(maybeConstructor: Option[ScConstructor]): Boolean = {
+    (for {
+      constructor <- maybeConstructor
+      ref <- constructor.reference
+    } yield {
+      ref.advancedResolve.filter(_.problems.isEmpty)
+        .map(_.element).exists {
+        case primary: ScPrimaryConstructor => true
+        case _ => false
+      }
+    }).getOrElse(true)
+  }
+
```

Comment:
Same thing, working with types is not necessary. You already do resolve in `isProblemlessPrimaryConstructor` method. So you may check that it's `containingClass` is case class. `ScPrimaryConstructor.ofClass` extractor can save you couple of lines here.